### PR TITLE
adopting s4

### PR DESCRIPTION
**Pull request by goloveychuk**

### CODE REVIEW

File: Source/Client.swift [hunk 1]
```diff
@@ -49,29 +49,28 @@ extension Client {
...
         }
 
         return request
     }
 
-    public func respond(request: Request) throws -> Response {
+    public func respond(to request: Request) throws -> Response {
         let request = addHeaders(request)
-        let stream = try client.connect()
+        try self.connection.open(timingOut: .never)
```
Comment (goloveychuk) [inline Source/Client.swift:61]: not sure if we should connect here every time. But it's the same behaviour as before.

File: Source/Client.swift [hunk 2]
```diff
@@ -49,29 +49,28 @@ extension Client {
...
-
-        try stream.flush()
+        try connection.flush()
 
         while true {
-            let data = try stream.receive()
+            var data = Data()
+            for chunk in StreamSequence(for: connection) {
+                data.bytes.append(contentsOf: chunk.bytes)
+            }
```
Comment (goloveychuk) [inline Source/Client.swift:70]: not beautiful

File: Source/Client.swift [hunk 3]
```diff
@@ -81,12 +80,12 @@ extension Client {
 
     public func send(request: Request, middleware: Middleware...) throws -> Response {
         let request = addHeaders(request)
-        return try middleware.intercept(self).respond(request)
+        return try middleware.chain(to: self).respond(to: request)
```
Comment (goloveychuk) [inline Source/Client.swift:83]: not sure if correctly replaced

File: Source/Client.swift (hunk 2 above)
Comment (dreymonde) [inline Source/Client.swift:70]:
``` swift
let data: Data = StreamSequence(for: connection).reduce([]) { $0.bytes.append(contentsOf: $1.bytes) }
```

File: Source/Client.swift (hunk 3 above)
Comment (dreymonde) [inline Source/Client.swift:83]: yeah, totally correct

File: Source/Client.swift (hunk 2 above)
Comment (Danappelxx) [inline Source/Client.swift:70]: While it isn't beautiful, the `reduce` @dreymonde is actually quadratic in time complexity (bytes array is copied every iteration). The for loop is a better approach.

A slightly more pretty alternative::

``` swift
let data: Data = {
    var data = Data()
    for chunk in StreamSequence(for: connection) {
        data.bytes.append(contentsOf: chunk.bytes)
    }
    return data
}()
```

**Pull request closed without merging**